src/RageSoundMixBuffer.cpp

```diff
@@ -31,19 +31,36 @@ void RageSoundMixBuffer::SetWriteOffset( int iOffset )
 	m_iOffset = iOffset;
 }
 
-void RageSoundMixBuffer::Extend( unsigned iSamples )
+void RageSoundMixBuffer::Extend(unsigned iSamples)
 {
-	const unsigned realsize = iSamples+m_iOffset;
-	if( m_iBufSize < realsize )
+	const std::uint64_t m_nBufSize = static_cast<std::uint64_t>(iSamples) + m_iOffset;
+	if (m_iBufSize < m_nBufSize)
 	{
-		m_pMixbuf = (float *) realloc( m_pMixbuf, sizeof(float) * realsize );
-		m_iBufSize = realsize;
+		m_pMixbufBackup = static_cast<float*>(realloc(m_pMixbufBackup, sizeof(float) * m_iBufSize));
+		if (m_pMixbufBackup != nullptr && m_pMixbuf != nullptr)
+		{
+			memcpy(m_pMixbufBackup, m_pMixbuf, m_iBufSize * sizeof(float));
+		}
+
+		float *m_pMixbufNew = static_cast<float*>(realloc(m_pMixbuf, sizeof(float) * m_nBufSize));
```

Comment:
The problem with the old code was apparently that `m_pMixbuf` might get reallocated while it was being read, but this new code does exactly the same thing here. It doesn't matter that you assign it to `m_pMixbufNew`, the old buffer will be invalid after this call (unless realloc fails).

Also similar note regarding `m_` prefix as above.

Reply:
It's only an issue if it is relocated while it's being read and then becomes NULL (fails to reallocate) for some reason. If that happens, it would lead to weird game stutters where the audio hangs too. With the backup buffer in place, even if the buffer glitches out the backup is there so it recovers quickly and gracefully. That's the important thing in the new code.

Reply:
If `realloc()` fails and returns `NULL`, the original buffer is kept intact. So the backup buffer doesn't seem to be necessary, we just need to keep the copy of the original pointer.

Reply:
What would you suggest changing in this case?

From my point of view, it's a very small amount memory increase to hold onto a backup buffer to ensure that even if something goes very wrong with the audio buffer, the game will recover gracefully. The difference is about a 25MB increase in RAM consumption on Windows between this code and the old code when letting a 10 song course play through on AutoPlay.

Reply:
The main reasons I'd like to keep the backup buffer system intact are

1. I've already tested this implementation thoroughly for reliability across multiple OS's and scenarios
2. It would be a lot of work to evaluate and prove whether or not holding the old pointer is at least as good as this system in terms of security
3. The memory or performance impact of maintaining the backup buffer is inconsequential (no notable increase in CPU usage or memory consumption with this code)